examples/mxnet/sampling/graphsage_cv.py

```diff
@@ -181,25 +178,33 @@ def forward(self, nf):
         return h
 
 
-def graphsage_cv_train(g, ctx, args, n_classes, train_nid, test_nid, n_test_samples):
+def graphsage_cv_train(g, ctx, args, n_classes, train_nid, test_nid, n_test_samples, distributed):
     features = g.ndata['features']
     labels = g.ndata['labels']
     in_feats = g.ndata['features'].shape[1]
+    g_ctx = features.context
 
     norm = mx.nd.expand_dims(1./g.in_degrees().astype('float32'), 1)
-    g.ndata['norm'] = norm.as_in_context(ctx)
-
+    g.ndata['norm'] = norm.as_in_context(g_ctx)
     degs = g.in_degrees().astype('float32').asnumpy()
     degs[degs > args.num_neighbors] = args.num_neighbors
-    g.ndata['subg_norm'] = mx.nd.expand_dims(mx.nd.array(1./degs, ctx=ctx), 1)
-
-    g.update_all(fn.copy_src(src='features', out='m'),
-                 fn.sum(msg='m', out='preprocess'),
-                 lambda node : {'preprocess': node.data['preprocess'] * node.data['norm']})
-
+    g.ndata['subg_norm'] = mx.nd.expand_dims(mx.nd.array(1./degs, ctx=g_ctx), 1)
     n_layers = args.n_layers
-    for i in range(n_layers):
-        g.ndata['h_{}'.format(i)] = mx.nd.zeros((features.shape[0], args.n_hidden), ctx=ctx)
+
+    if distributed:
+        g.dist_update_all(fn.copy_src(src='features', out='m'),
+                          fn.sum(msg='m', out='preprocess'),
+                          lambda node : {'preprocess': node.data['preprocess'] * node.data['norm']})
+        for i in range(n_layers):
+            g.init_ndata('h_{}'.format(i), (features.shape[0], args.n_hidden), 'float32')
+            g.init_ndata('agg_h_{}'.format(i), (features.shape[0], args.n_hidden), 'float32')
```

Comment:
ditto

Reply:
I actually try to explain these two new APIs in the tutorial: https://hackmd.io/szg0gaArQg-DVbo1_JjTcA?both#